src/index.js

```diff
@@ -1,390 +1,648 @@
-export function h(name, attributes) {
-  var rest = []
-  var children = []
-  var length = arguments.length
+var DEFAULT = 0
+var RECYCLED_NODE = 1
+var LAZY_NODE = 2
+var TEXT_NODE = 3 // Node.TEXT_NODE
 
-  while (length-- > 2) rest.push(arguments[length])
+var XLINK_NS = "http://www.w3.org/1999/xlink"
+var SVG_NS = "http://www.w3.org/2000/svg"
 
-  while (rest.length) {
-    var node = rest.pop()
-    if (node && node.pop) {
-      for (length = node.length; length--; ) {
-        rest.push(node[length])
-      }
-    } else if (node != null && node !== true && node !== false) {
-      children.push(node)
-    }
-  }
+var EMPTY_OBJECT = {}
+var EMPTY_ARRAY = []
 
-  return typeof name === "function"
-    ? name(attributes || {}, children)
-    : {
-        nodeName: name,
-        attributes: attributes || {},
-        children: children,
-        key: attributes && attributes.key
-      }
-}
+var map = EMPTY_ARRAY.map
+var isArray = Array.isArray
 
-export function app(state, actions, view, container) {
-  var map = [].map
-  var rootElement = (container && container.children[0]) || null
-  var oldNode = rootElement && recycleElement(rootElement)
-  var lifecycle = []
-  var skipRender
-  var isRecycling = true
-  var globalState = clone(state)
-  var wiredActions = wireStateToActions([], globalState, clone(actions))
-
-  scheduleRender()
-
-  return wiredActions
-
-  function recycleElement(element) {
-    return {
-      nodeName: element.nodeName.toLowerCase(),
-      attributes: {},
-      children: map.call(element.childNodes, function(element) {
-        return element.nodeType === 3 // Node.TEXT_NODE
-          ? element.nodeValue
-          : recycleElement(element)
-      })
-    }
-  }
+var merge = function(a, b) {
+  var target = {}
 
-  function resolveNode(node) {
-    return typeof node === "function"
-      ? resolveNode(node(globalState, wiredActions))
-      : node != null
-        ? node
-        : ""
-  }
+  for (var i in a) target[i] = a[i]
+  for (var i in b) target[i] = b[i]
 
-  function render() {
-    skipRender = !skipRender
+  return target
+}
 
-    var node = resolveNode(view)
+var resolved = typeof Promise === "function" && Promise.resolve()
 
-    if (container && !skipRender) {
-      rootElement = patch(container, rootElement, oldNode, (oldNode = node))
+var defer = !resolved
+  ? setTimeout
+  : function(cb) {
+      return resolved.then(cb)
```

Comment:
If I understand correctly this function produce an infinite promise chain, is it a memory leak?

Reply:
Not quite, `resolved` is never updated, so it's essentially forcing js to wait a single tick. The alternative is something like `setTimeout(cb, 0)` which I believe isn't as fast as a quick-resolve like this.

Reply:
Yeah. `setTimeout` schedules a task, while our `defer` schedules a "micro" task, which has more priority. All this is untested science. I don't if it'd make a difference to use `setTimeout` instead. I can remove this until we've seen positive benchmark results if you want.

Reply:
For what it's worth, `ivi` also uses `Promise.resolve` to schedule a microtask. 

https://github.com/localvoid/ivi/blob/cdaf62c521b1d32c601ab5ed0a0cd226ed5921c5/packages/ivi-scheduler/src/index.ts#L1-L7

Reply:
Yes, but they create a new `Promise.resolve()` every tick and not use `.then()` for the same resolved promise up to infinity.

I thought it may cause a memory leak and created a small demo with a try to reproduce it: https://frenzzy.github.io/test/then.html (or [codepen](https://codepen.io/frenzzy/pen/gEPdRa?editors=1010))
But looks like there are no problem at least in the latest Chrome.

Reply:
@frenzzy  I just tried your example in IE11 with a Promise polyfill, and there was no memory leak.

Reply:
@SkaterDad `ivi` is using [its own microtask queue](https://github.com/localvoid/ivi/blob/cdaf62c521b1d32c601ab5ed0a0cd226ed5921c5/packages/ivi/src/scheduler/index.ts#L91) and `Promise.resolve` is invoked only when microtask is added [outside of a scheduler context](https://github.com/localvoid/ivi/blob/cdaf62c521b1d32c601ab5ed0a0cd226ed5921c5/packages/ivi/src/scheduler/index.ts#L131). There are edge cases when some DOM operations are blocked when they executed outside of an event context, that is one of the reasons why React and ivi executes event handlers inside of a scheduler context.

Reply:
Okay, folks, I changed it to:

```js
var defer =
  typeof Promise === "function"
    ? function(cb) {
        Promise.resolve().then(cb)
      }
    : setTimeout
```

Reply:
Okay, folks, I changed it to:

```js
function defer(cb) {
  Promise.resolve().then(cb)
}
```